Add `examples` field to `ToolDefinition` and send it to Anthropic

AlanPonnachan opened the pull request:
- Closes #4168

Adds structured tool input examples across `ToolDefinition`, function tools, agent decorators, function toolsets, and `ToolOutput`.

Anthropic transports that support `input_examples` receive examples natively. Legacy Amazon Bedrock and other providers receive a JSON fallback in the tool description. Function-tool examples are caller-authored wire objects, while `ToolOutput` examples are semantic output values serialized into the output tool input shape.

Includes public API compatibility tests, transport-specific profile coverage, an end-to-end Anthropic cassette assertion, fallback tests, documentation, and bundled agent guidance.

### Checklist

- [ ] Any **AI generated code** has been reviewed line-by-line by the human PR author, who stands by it.
- [x] No **breaking changes** in accordance with the [version policy](https://github.com/pydantic/pydantic-ai/blob/main/docs/version-policy.md).
- [x] **PR title** is fit for the [release changelog](https://github.com/pydantic/pydantic-ai/releases).

AlanPonnachan committed:
add tool examples to anthropic

AlanPonnachan committed:
add flattening

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

Kludex reviewed (changes requested):
I think we should be able to pass from the docstring, since Google docstring style has the `Examples:`, e.g.: https://gist.github.com/redlotus/3bc387c2591e3e908c9b63b97b11d24e#file-docstrings-py-L142-L148

AlanPonnachan committed:
add ability to pass docstring

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

dsfaccini commented on pydantic_ai_slim/pydantic_ai/tools.py:
```diff
@@ -516,6 +530,18 @@ class ToolDefinition:
+    examples: list[dict[str, Any]] | None = None
+    """Example inputs demonstrating correct tool usage patterns.
+
+    Provide 1-5 realistic examples showing parameter conventions, optional field patterns,
```
where does the `1-5` come from? if this is anthropic specific it should have a respective profile setting (e.g. `max_nr_of_examples_supported`

AlanPonnachan committed:
apply suggestions

DouweM committed:
Merge branch 'main' into feat/anthropic-tool-examples

DouweM commented on pydantic_ai_slim/pydantic_ai/_griffe.py:
```diff
@@ -65,6 +65,22 @@ def doc_descriptions(
         main_desc = main.value
 
+    # Handle Examples section
+    if examples := next((p for p in sections if p.kind == DocstringSectionKind.examples), None):
```
What's all this for? We don't seem to have a test that does anything with docstrings

DouweM commented on pydantic_ai_slim/pydantic_ai/_output.py:
```diff
@@ -904,6 +906,10 @@ def build(
             processor = ObjectOutputProcessor(output=output, description=description, strict=strict)
+            # Flatten examples if the output function has a single argument that gets flattened in the schema
+            if examples and processor._function_schema and processor._function_schema.single_arg_name:  # pyright: ignore[reportPrivateUsage]
+                arg_name = processor._function_schema.single_arg_name  # pyright: ignore[reportPrivateUsage]
```
- It think we should do something similar for `processor.outer_typed_dict_key` (which is populated for e.g. `output_type=float`): wrapping the example in it
- Can we do this without using private properties? The processor could have a public property
- We're reading the field two times

DouweM commented on pydantic_ai_slim/pydantic_ai/output.py:
```diff
@@ -125,12 +127,14 @@ def __init__(
         description: str | None = None,
         max_retries: int | None = None,
         strict: bool | None = None,
+        examples: list[dict[str, Any]] | None = None,
```
Specifically on `ToolOutput`, I think this should be `list[Any]`, as you could have `output_type=ToolOutput(list[Foo], examples=[Foo(1), Foo(2), Foo(3)])`. That's where wrapping in `outer_typed_dict_key` comes in, as I mentioned above.

DouweM commented on pydantic_ai_slim/pydantic_ai/tools.py:
```diff
@@ -378,6 +382,11 @@ async def prep_my_tool(
+        # Flatten examples if the tool has a single argument that gets flattened in the schema
+        if self.examples and self.function_schema.single_arg_name:
+            arg_name = self.function_schema.single_arg_name
+            self.examples = [ex[arg_name] if isinstance(ex, dict) and arg_name in ex else ex for ex in self.examples]
```
Can we somehow not duplicate this with the output code?

DouweM commented on tests/models/test_anthropic.py:
```diff
@@ -8293,3 +8302,73 @@ async def test_anthropic_system_prompts_and_instructions_ordering():
+    m = AnthropicModel('claude-3-5-sonnet-latest', provider=AnthropicProvider(anthropic_client=mock_client))
+
+    from pydantic_ai.models import ModelRequestParameters
+    from pydantic_ai.models.anthropic import AnthropicModelSettings
```
Imports only at the top of the file please

DouweM commented on tests/test_tools.py:
```diff
@@ -2776,3 +2793,99 @@ def test_agent_tool_timeout_passed_to_toolset():
+    # Access the output toolset to verify definition
+    assert agent._output_toolset is not None
+    tool_def = agent._output_toolset._tool_defs[0]
+    assert tool_def.examples == examples
```
This example is invalid! `{'x': 1}` does not actually match the schema, which would require `{'result': int}` because of the `outer_typed_dict_key` I mentioned above. So really we should pass `examples=[1]` above, and then here we should see `[{'result': 1}]`

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

dsfaccini committed:
Merge branch 'main' into feat/anthropic-tool-examples

AlanPonnachan committed:
resolves issues and fix tests

AlanPonnachan committed:
fix issues

AlanPonnachan committed:
test coverage

AlanPonnachan committed:
test coverage

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

DouweM commented on pydantic_ai_slim/pydantic_ai/tools.py:
```diff
@@ -428,18 +437,42 @@ def from_schema(
+            if description:
+                description = f'{description}\n\nExamples:\n{examples_str}'
+            else:
+                description = f'Examples:\n{examples_str}'
```
Right, we should make sure we only add them to the description in `Model.prepare_request`, once it's been verified that the model profile does not natively support this feature. We're doing the same thing for `return_schema` (which Google supports natively) in https://github.com/pydantic/pydantic-ai/pull/3865

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

DouweM commented:
@AlanPonnachan That all makes sense to me! @adtyavrdhn Can you have a look here please as this is similar to #3865, where we're sending the return schema in the description or a Google API field?

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

AlanPonnachan committed:
update pr with new changes

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

AlanPonnachan commented on pydantic_ai_slim/pydantic_ai/models/anthropic.py:
```diff
@@ -1147,6 +1147,9 @@ def _map_tool_definition(self, f: ToolDefinition) -> BetaToolParam:
             tool_param['strict'] = f.strict
+
+        if f.examples and self.profile.supports_tool_examples:
+            tool_param['input_examples'] = f.examples
```
`input_examples` is now natively supported without beta headers for the latest models. Update PR description

AlanPonnachan committed:
setting examples=None on replace call

DouweM committed:
Add test cassette

AlanPonnachan commented on pydantic_ai_slim/pydantic_ai/profiles/anthropic.py:
```diff
@@ -17,8 +17,13 @@ def anthropic_model_profile(model_name: str) -> ModelProfile | None:
     # TODO update when new models are released that support structured outputs
     # https://docs.claude.com/en/docs/build-with-claude/structured-outputs#example-usage
 
+    models_that_support_tool_examples = ('claude-opus-4-6',)
```
In docs, there is no explicit mention of supported models.
As of now i given 
('claude-sonnet-4-5', 'claude-sonnet-4-6', 'claude-opus-4-5', 'claude-opus-4-6')

AlanPonnachan committed:
resolve review points

AlanPonnachan committed:
add documentation

AlanPonnachan committed:
Merge main and resolve uv.lock

AlanPonnachan committed:
test coverage

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

AlanPonnachan committed:
trigger ci rerun

AlanPonnachan committed:
test coverage

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

AlanPonnachan committed:
add missing examples definition

AlanPonnachan committed:
support tool examples correction

AlanPonnachan committed:
test coverage

adtyavrdhn commented:
Sorry about the delay, I missed this notification 😞 

This looks inline with what we did for Tool `return_schema`. The design is sound.

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

AlanPonnachan commented:
conflicts resolved, will check the ci and update

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

DouweM committed:
Merge branch 'main' into feat/anthropic-tool-examples

DouweM commented:
@AlanPonnachan Appreciate it Alan, it's on my list to get back to next week.

DouweM commented:
@AlanPonnachan Any chance you can get CI green for me? 😇

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

AlanPonnachan committed:
resolve ci checks

AlanPonnachan committed:
fix issues

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

AlanPonnachan commented:
@DouweM Sorry for the delay, it’s been a busy week. CI is green now.

DouweM committed:
Merge remote-tracking branch 'origin/main' into feat/anthropic-tool-examples

# Conflicts:
#	pydantic_ai_slim/pydantic_ai/providers/bedrock.py

DouweM committed:
Move tool-examples support flag into Anthropic model profile

Set  in  alongside the other
model-name-gated capability flags (matching ), and
add  to the allowlist. Both Bedrock paths override it to False:
the Converse `toolSpec` has no `input_examples` field, and the Anthropic-on-Bedrock
client would need the `tool-examples-2025-10-29` beta header we don't send.

DouweM commented:
I'm about to go on parental leave (back on June 29), I'll continue on this when I'm back, appreciate the patience!

AlanPonnachan committed:
Merge branch 'main' into feat/anthropic-tool-examples

adtyavrdhn committed:
Merge remote-tracking branch 'origin/main' into feat/anthropic-tool-examples

# Conflicts:
#	docs/output.md
#	pydantic_ai_slim/pydantic_ai/tools.py

adtyavrdhn committed:
Fix tool example API compatibility and Anthropic transports

adtyavrdhn committed:
Validate structured tool examples against schemas

adtyavrdhn committed:
Preserve schema-facing tool example inputs

adtyavrdhn committed:
Keep tool examples as caller-authored wire inputs

adtyavrdhn committed:
Clarify tool example input contracts

adtyavrdhn committed:
Update tool example profile snapshots

adtyavrdhn committed:
Exclude uncalled validator from coverage

adtyavrdhn committed:
Exercise positional validator coverage

adtyavrdhn committed:
Move tool example serialization to shared utilities

adtyavrdhn commented on pydantic_ai_slim/pydantic_ai/providers/anthropic.py:
```diff
@@ -2,7 +2,7 @@
 import os
 import warnings
-from dataclasses import dataclass
+from dataclasses import dataclass, replace
```
Valid and fixed. The provider no longer calls `dataclasses.replace()` on a `TypedDict`; profile overrides use `merge_profile()` and `AnthropicModelProfile(...)`.

adtyavrdhn commented on tests/models/test_anthropic.py:
```diff
@@ -8292,3 +8292,83 @@ async def test_anthropic_system_prompts_and_instructions_ordering():
+    completion_kwargs = get_mock_chat_completion_kwargs(mock_client)[0]
+    betas = completion_kwargs['betas']
+    assert 'advanced-tool-use-2025-11-20' in betas
+    assert 'tool-examples-2025-10-29' not in betas
```
Valid for the earlier beta API, but superseded by Anthropic making `input_examples` a standard tool field. The current VCR test asserts `input_examples` in the wire payload, and transport tests cover native support versus legacy Bedrock fallback.

adtyavrdhn commented on pydantic_ai_slim/pydantic_ai/output.py:
```diff
@@ -121,6 +121,8 @@ class Vehicle(BaseModel):
     strict: bool | None
     """Whether to use strict mode for the tool."""
+    examples: list[Any] | None
+    """Example inputs demonstrating correct tool usage."""
```
The dedicated output docs now explain the `ToolOutput` contract in detail. We intentionally did not cross-reference `ToolDefinition.examples` from the attribute because `ToolOutput` accepts semantic output values, while `ToolDefinition` holds serialized JSON object inputs; conflating those contracts would be misleading.

adtyavrdhn commented on pydantic_ai_slim/pydantic_ai/providers/anthropic.py:
```diff
@@ -62,7 +62,19 @@ def model_profile(model_name: str) -> ModelProfile | None:
+            'claude-opus-4-5',
+            'claude-opus-4-6',
+        )
+        supports_tool_examples = model_name.startswith(models_that_support_tool_examples)
```
Valid. The stale model allowlist was removed. Anthropic provider profiles now enable standard tool examples without enumerating model IDs, including current and future models, with explicit coverage for a future-style model ID.

adtyavrdhn commented on pydantic_ai_slim/pydantic_ai/_griffe.py:
```diff
@@ -65,18 +66,38 @@ def doc_descriptions(
+    if examples := next((p for p in sections if p.kind == DocstringSectionKind.examples), None):
+        examples_content: list[str] = []
+        for source, output in examples.value:
+            if isinstance(source, DocstringSectionKind):  # pragma: no branch
```
The Griffe examples-section change was removed from this PR, including this branch pragma. Structured tool examples are now supplied only through the explicit `examples` API.

adtyavrdhn commented on pydantic_ai_slim/pydantic_ai/_output.py:
```diff
@@ -32,7 +32,12 @@
+    GenerateToolJsonSchema,
+    ObjectJsonSchema,
+    ToolDefinition,
+    _process_examples,  # pyright: ignore[reportPrivateUsage]
```
Valid and fixed in `a4267c5d9`. Serialization now lives in `_utils.process_tool_examples`; both tool and output paths use the shared private module, and the cross-module private import plus Pyright suppression are gone.

adtyavrdhn commented on tests/models/test_anthropic.py:
```diff
@@ -8409,3 +8415,28 @@ async def test_anthropic_system_prompts_and_instructions_ordering():
+    assert mapped.get('input_examples') == examples
+
+
+@pytest.mark.vcr()
```
The cassette has been recorded and replaced by the current `claude-opus-4-6` VCR coverage, which asserts `input_examples` in each recorded request body.

adtyavrdhn commented on pydantic_ai_slim/pydantic_ai/_griffe.py:
```diff
@@ -65,6 +65,22 @@ def doc_descriptions(
+                examples_content.append(source)
+
+        if examples_content:
+            formatted_examples = '\n\n'.join(examples_content)
```
The docstring parsing change was removed from this PR to keep the feature scoped. Native `input_examples` now comes only from the explicit structured `examples` API, so there is no brittle conversion from prose or doctest content.

adtyavrdhn commented on pydantic_ai_slim/pydantic_ai/models/__init__.py:
```diff
@@ -340,6 +340,14 @@ def prepare_request(
+                params,
+                function_tools=[self._append_tool_examples(t) for t in params.function_tools],
+                output_tools=[self._append_tool_examples(t) for t in params.output_tools],
+            )
```
Valid and fixed. All profile reads use dictionary access with `.get('supports_tool_examples', False)`, with runtime and type-check coverage.

adtyavrdhn commented on pydantic_ai_slim/pydantic_ai/profiles/__init__.py:
```diff
@@ -108,6 +108,9 @@ class ModelProfile(TypedDict, total=False):
 
+    supports_tool_examples: bool = False
+    """Whether the model natively supports providing structured tool examples."""
+
```
Valid and fixed. `supports_tool_examples` is now a bare optional `TypedDict` item, with defaults handled at profile merge and access sites.

adtyavrdhn commented on docs/output.md:
```diff
@@ -357,6 +357,36 @@ print(repr(result.output))
+You can provide examples of correct tool usage to improve the model's understanding of your tool's expected input. Examples can be provided in two ways:
+
+1. **Structured Examples**: Pass a list of examples to the `examples` parameter in [`@agent.tool`][pydantic_ai.agent.Agent.tool], [`@agent.tool_plain`][pydantic_ai.agent.Agent.tool_plain], or the [`Tool`][pydantic_ai.tools.Tool] class directly.
+2. **Docstring Examples**: Include an `Examples:` section in the tool's docstring.
```
Valid and fixed. The output documentation now uses `ToolOutput(..., examples=...)`, explains semantic output values, and uses the distinct `output_examples.py` title.

adtyavrdhn commented on pydantic_ai_slim/pydantic_ai/providers/anthropic.py:
```diff
@@ -62,7 +62,13 @@ def model_profile(model_name: str) -> ModelProfile | None:
+            # `anthropic_model_profile` enables `supports_tool_examples` for the direct Anthropic API, where
+            # `input_examples` is a standard tool field. Via Bedrock it's still beta-gated behind the
+            # `tool-examples-2025-10-29` header (which we don't send), so fall back to the description instead.
+            result = replace(result, supports_tool_examples=False)
```
Valid concern and addressed with client-based transport gating. Direct Anthropic, Vertex, Foundry, and Bedrock Mantle use the standard `input_examples` field; legacy `AsyncAnthropicBedrock` falls back to descriptions. Transport tests cover the legacy and Mantle distinction, and the direct VCR asserts the wire payload.

adtyavrdhn commented on docs/tools.md:
````diff
@@ -375,6 +375,36 @@ _(This example is complete, it can be run "as is")_
+```python {title="tool_examples.py"}
+from pydantic_ai import Agent
+
+agent = Agent('anthropic:claude-opus-4-6')
````
Not applicable to the current docs test setup. The documentation suite installs all provider extras, this snippet constructs but does not run the Anthropic model, and the example matrix passes on Python 3.11 through 3.14 without credentials.

adtyavrdhn commented on pydantic_ai_slim/pydantic_ai/toolsets/function.py:
```diff
@@ -188,6 +189,7 @@ def tool(
         strict: bool | None = None,
         sequential: bool | None = None,
         requires_approval: bool | None = None,
+        examples: list[Any] | None = None,
```
Valid and fixed. `FunctionToolset.tool()` now forwards `examples` to `add_function()`, with a regression test covering decorator forwarding and positional API compatibility.

DouweM commented:
Closed, see issue.
